evap/rewards/tools.py

```diff
@@ -71,14 +84,21 @@ def grant_reward_points(user, semester):
     required_courses = Course.objects.filter(participants=user, semester=semester, is_required_for_reward=True)
     if not required_courses.exists():
         return False
-    # does the user not participate in any more required courses in this semester?
-    if required_courses.exclude(voters=user).exists():
-        return False
-    # did the user not already get reward points for this semester?
-    if RewardPointGranting.objects.filter(user_profile=user, semester=semester).exists():
+
+    voted_courses_count = required_courses.filter(voters=user).count()
+    # full evaluation progress from 0.0 to 1.0
+    progress = float(voted_courses_count) / float(required_courses.count())
+
+    # How many points have been granted to this user this semester?
+    granted_points = RewardPointGranting.objects.filter(user_profile=user, semester=semester).aggregate(Sum('value'))['value__sum'] or 0
+    points_missing = target_points(progress) - granted_points
+
+    # is the user actually missing points?
+    if points_missing < 1:
         return False
-    # grant reward points
-    RewardPointGranting.objects.create(user_profile=user, semester=semester, value=settings.REWARD_POINTS_PER_SEMESTER)
+
+    # grant missing reward points
+    RewardPointGranting.objects.create(user_profile=user, semester=semester, value=points_missing)
     return True
```

Comment:
also, i would switch that if around (and maybe even inline points_missing):
```
if target_points > granted_points:
    RewardPointGranting.objects.create(..., value=target_points-granted_points)
    return True
return False
```
for me that expresses exactly what needs to be done, no further comments needed. but if you want to keep the early `return False` to be in line with the rest, that's fine as well.